src/ipc/ipc4/mtrace.c

```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Kai Vehmanen <kai.vehmanen@linux.intel.com>
+
+#include <stdint.h>
+#include <sof/lib/uuid.h>
+#include <sof/ipc/common.h>
+#include <ipc4/base_fw.h>
+#include <ipc4/error_status.h>
+
+#ifdef CONFIG_LOG_BACKEND_ADSP_MTRACE
+
+#undef panic // SOF headers may redefine panic
+#include <zephyr/logging/log_backend.h>
+#include <zephyr/logging/log_backend_adsp_mtrace.h>
+#include <zephyr/logging/log.h>
+
+#include <sof/schedule/ll_schedule.h>
+#include <sof/schedule/schedule.h>
+#include <sof/schedule/task.h>
+
+LOG_MODULE_REGISTER(mtrace, CONFIG_SOF_LOG_LEVEL);
+
+/**
+ * If mtrace log buffer has less space than
+ * the threshold, notify host with BUFFER_STATUS IPC4.
+ */
+#define NOTIFY_BUFFER_STATUS_THRESHOLD 2048
+
+/**
+ * Default aging timer value. This time defines the maximum time to
+ * block IPC BUFFER STATUS notify messages. The notification is sent
+ * either due to amount of data available in the buffer, or by timeout
+ * in case logs are sent at slow rate.
+ */
+#define IPC4_MTRACE_NOTIFY_AGING_TIMER_MS 1000
+
+/**
+ * Smallest possible aging timer value
+ */
+#define IPC4_MTRACE_AGING_TIMER_MIN_MS 100
+
+/* bb2aa22e-1ab6-4650-8501-6e67fcc04f4e */
+DECLARE_SOF_UUID("mtrace-task", mtrace_task_uuid, 0xbb2aa22e, 0x1ab6, 0x4650,
+		 0x85, 0x01, 0x6e, 0x67, 0xfc, 0xc0, 0x4f, 0x4e);
+
+static uint64_t mtrace_notify_last_sent;
+
+static uint32_t mtrace_bytes_pending;
+
+static uint32_t mtrace_aging_timer = IPC4_MTRACE_NOTIFY_AGING_TIMER_MS;
+
+static struct k_thread mtrace_thread;
+
+#define MTRACE_STACK_SIZE	4096
+#define MTRACE_IPC_CORE	PLATFORM_PRIMARY_CORE_ID
+#define MTRACE_THREAD_PRIO      -1
+#define MTRACE_THREAD_NAME      "mtrace"
+
+K_KERNEL_STACK_DEFINE(mtrace_stack, MTRACE_STACK_SIZE);
+
+static void mtrace_log_hook(size_t written, size_t space_left)
+{
+	if (arch_proc_id() == MTRACE_IPC_CORE &&
+	    (space_left < NOTIFY_BUFFER_STATUS_THRESHOLD ||
+	     k_uptime_delta(&mtrace_notify_last_sent) >= mtrace_aging_timer)) {
+		ipc_send_buffer_status_notify();
+		mtrace_notify_last_sent = k_uptime_get();
```

Comment:
looks like locking is needed. What if the logging thread is run and has already notified the host, but here it's interrupted by an interrupt, which logs too and increments `mtrace_bytes_pending` after that we return to the thread, and it sets `mtrace_bytes_pending = 0`

Reply:
Ack, let me look at this. The mtrace_bytes_pending value is actually not sent to the host. The IPC just tells host that "there's something in the buffer" and host will go and check the read/write pointers directly from SRAM. So I think in this case it's enough we send one indication and host will also retrieve the logs written from interrupt. And I'm a bit worried about using spinlocks here as the log callbacks may come from e.g. panic handler.

Reply:
right, but `mtrace_bytes_pending` is checked below in the thread, so if it's 0, the host will not be notified. Sure, you cannot spinlock around `ipc_send_buffer_status_notify()` but maybe something like this would work:
```
size_t bytes_reported = mtrace_bytes_pending;
ipc_send_buffer_status_notify();
spin_lock();
mtrace_bytes_pending -= bytes_reported;
spin_unlock();
```
and below you'd hold the same spinlock when adding to `mtrace_bytes_pending`

Reply:
@lyakh I checked and the logging backend is serializing the hook calls, so the above race cannot handle.